sources/Outputs/src/Par.cpp

```diff
@@ -0,0 +1,174 @@
+  set->addReference(helper::buildReference("generator_PNom", "p_pu", "DOUBLE"));
+  set->addReference(helper::buildReference("generator_P0Pu", "p_pu", "DOUBLE"));
+  set->addReference(helper::buildReference("generator_Q0Pu", "q_pu", "DOUBLE"));
+  set->addReference(helper::buildReference("generator_U0Pu", "v_pu", "DOUBLE"));
+  set->addReference(helper::buildReference("generator_UPhase0", "angle_pu", "DOUBLE"));
+}
+
+std::vector<boost::shared_ptr<parameters::ParametersSet>>
+Par::writeConstantSets(unsigned int nb_generators) {
+  std::vector<boost::shared_ptr<parameters::ParametersSet>> ret;
+
+  // Load
+  auto set = parameters::ParametersSetFactory::newInstance(constants::loadParId);
+  set->addParameter(helper::buildParameter("load_Alpha", 1.5));
+  set->addParameter(helper::buildParameter("load_Beta", 2.5));
+  set->addParameter(helper::buildParameter("load_UMaxPu", 1.05));
+  set->addParameter(helper::buildParameter("load_UMinPu", 0.95));
+  set->addParameter(helper::buildParameter("load_tFilter", 10.));
+  set->addReference(helper::buildReference("load_P0Pu", "p_pu", "DOUBLE"));
+  set->addReference(helper::buildReference("load_Q0Pu", "q_pu", "DOUBLE"));
+  set->addReference(helper::buildReference("load_U0Pu", "v_pu", "DOUBLE"));
+  set->addReference(helper::buildReference("load_UPhase0", "angle_pu", "DOUBLE"));
+
+  ret.push_back(set);
+
+  // Signal N
+  set = parameters::ParametersSetFactory::newInstance(constants::signalNParId);
+  set->addParameter(helper::buildParameter("nbGen", static_cast<int>(nb_generators)));  // static cast because of Dynawo API
+  ret.push_back(set);
+
+  // Signal N generator
+  set = parameters::ParametersSetFactory::newInstance(constants::signalNGeneratorParId);
+  updateSignalNGenerator(set);
+  ret.push_back(set);
+
+  // Signal N generator with impendance
+  set = parameters::ParametersSetFactory::newInstance(constants::impSignalNGeneratorParId);
+  updateSignalNGenerator(set);
+  updateCouplingParameters(set);
+  ret.push_back(set);
+
+  return ret;
+}
+
+boost::shared_ptr<parameters::ParametersSet>
+Par::writeGenerator(const algo::GeneratorDefinition& def, const std::string& basename) {
+  auto set = parameters::ParametersSetFactory::newInstance(def.id);
+
+  if (def.model == algo::GeneratorDefinition::ModelType::SIGNALN || def.model == algo::GeneratorDefinition::ModelType::WITH_IMPEDANCE_SIGNALN) {
+    // already processed by constant
+    return nullptr;
+  }
+
+  set->addParameter(helper::buildParameter("generator_KGover", 1.));
+  if (def.model == algo::GeneratorDefinition::ModelType::WITH_IMPEDANCE_DIAGRAM_PQ_SIGNALN ||
+      def.model == algo::GeneratorDefinition::ModelType::WITH_IMPEDANCE_SIGNALN) {
+    updateCouplingParameters(set);
+  }
+
+  set->addParameter(helper::buildParameter("generator_tFilter", 0.001));
+  // Qmax and QMin are determined in dynawo according to reactive capabilities curves and min max
+  set->addParameter(helper::buildParameter("generator_QMin0", def.qmin - 1));
```

Comment:
could you add a comment here about def.qmin -1 and def.qmax + 1, so that we do not forget why we do that? the answer is that we need a small numerical tolerance in case the starting point of the reactive injection is exactly on the limit of the reactive capability curve

Reply:
OK